File: calculator.py
"""Liquidity calculation engine matching the spreadsheet logic."""

from config import DEFAULT_DEPTH_PERCENT, DEFAULT_THRESHOLD


def calculate_liquidity(bids: list[dict], depth_percent: float = DEFAULT_DEPTH_PERCENT,
                        custom_vol: float | None = None,
                        threshold: float = DEFAULT_THRESHOLD) -> dict:
    if not bids:
        return _empty_result(threshold)

    # Build order book table
    levels = []
    accru_amount = 0.0
    accru_thb = 0.0

    for bid in bids:
        price = bid["price"]
        amount = bid["amount"]
        bid_size = amount * price
        accru_amount += amount
        accru_thb += bid_size
        levels.append({
            "price": price,
            "amount": amount,
            "bid_size": bid_size,
            "accru_amount": accru_amount,
            "accru_thb": accru_thb,
        })

    total_amount = accru_amount
    best_bid = levels[0]["price"]
    worst_bid = levels[-1]["price"]

    # Determine vol_used
    if custom_vol is not None:
        vol_used = min(custom_vol, total_amount)
    else:
        vol_used = total_amount * depth_percent

    # Walk through bids to fill vol_used
    matched_levels = _fill_orders(levels, vol_used)
    vol_received = sum(ml["sales_matched"] for ml in matched_levels)

    # Accumulated matched THB
    accru_matched = 0.0
    for ml in matched_levels:
        accru_matched += ml["sales_matched"]
        ml["accru_matched"] = accru_matched

    # Slippage
    expected_thb = vol_used * best_bid
    diff = vol_received - expected_thb
    slippage = diff / expected_thb if expected_thb > 0 else 0.0

    # Safety line
    safety = _calculate_safety_line(levels, best_bid, threshold)

    return {
        "best_bid": best_bid,
        "worst_bid": worst_bid,
        "total_amount": total_amount,
        "total_thb": levels[-1]["accru_thb"] if levels else 0.0,
        "vol_used": vol_used,
        "vol_received": vol_received,
        "diff": diff,
        "slippage": slippage,
        "slippage_pct": f"{slippage * 100:.3f}%",
        "threshold": threshold,
        "threshold_breached": slippage < threshold,
        "safety": safety,
        "levels": matched_levels,
    }


def _fill_orders(levels: list[dict], vol_target: float) -> list[dict]:
    matched = []
    remaining = vol_target
    for level in levels:
        if remaining <= 0:
            matched.append({**level, "amount_match": 0.0, "sales_matched": 0.0})
            continue
        match_amount = min(remaining, level["amount"])
        match_thb = match_amount * level["price"]
        remaining -= match_amount
        matched.append({**level, "amount_match": match_amount, "sales_matched": match_thb})
    return matched


def _calculate_safety_line(levels: list[dict], best_bid: float,
                           threshold: float) -> dict:
    if best_bid <= 0 or not levels:
        return {"safe_vol": 0, "safe_thb": 0, "crossed_at_level": -1,
                "is_safe": False}

    accru_vol = 0.0
    accru_thb = 0.0

    for i, level in enumerate(levels):
        price = level["price"]
        amount = level["amount"]

        next_vol = accru_vol + amount
        next_thb = accru_thb + amount * price
        expected = next_vol * best_bid
        slip = (next_thb - expected) / expected if expected > 0 else 0

        if slip < threshold:
            # Interpolate exact crossing point
            t = threshold
            numerator = (1 + t) * accru_vol * best_bid - accru_thb
            denominator = price - (1 + t) * best_bid
            if denominator != 0:
                x = max(0, min(numerator / denominator, amount))
            else:
                x = 0
            safe_vol = accru_vol + x
            safe_thb = accru_thb + x * price
            return {
                "safe_vol": safe_vol,
                "safe_thb": safe_thb,
                "crossed_at_level": i,
                "is_safe": False,
            }

        accru_vol = next_vol
        accru_thb = next_thb

    # Never crossed threshold - entire book is safe
    return {
        "safe_vol": accru_vol,
        "safe_thb": accru_thb,
        "crossed_at_level": -1,
        "is_safe": True,
    }


def _empty_result(threshold: float = DEFAULT_THRESHOLD) -> dict:
    return {
        "best_bid": 0, "worst_bid": 0, "total_amount": 0, "total_thb": 0,
        "vol_used": 0, "vol_received": 0, "diff": 0, "slippage": 0,
        "slippage_pct": "N/A", "threshold": threshold,
        "threshold_breached": False,
        "safety": {"safe_vol": 0, "safe_thb": 0, "crossed_at_level": -1, "is_safe": False},
        "levels": [],
    }

File: src/lib/calculator.js
import { DEFAULT_DEPTH, DEFAULT_THRESHOLD } from './config.js';

export function calculateLiquidity(bids, depthPercent = DEFAULT_DEPTH, customVol = null, threshold = DEFAULT_THRESHOLD) {
  if (!bids || !bids.length) return emptyResult(threshold);

  // Build levels
  const levels = [];
  let accruAmount = 0, accruThb = 0;
  for (const bid of bids) {
    const bidSize = bid.amount * bid.price;
    accruAmount += bid.amount;
    accruThb += bidSize;
    levels.push({ price: bid.price, amount: bid.amount, bid_size: bidSize, accru_amount: accruAmount, accru_thb: accruThb });
  }

  const totalAmount = accruAmount;
  const bestBid = levels[0].price;
  const worstBid = levels[levels.length - 1].price;

  const volUsed = customVol != null ? Math.min(customVol, totalAmount) : totalAmount * depthPercent;

  // Fill orders
  const matchedLevels = fillOrders(levels, volUsed);
  const volReceived = matchedLevels.reduce((s, l) => s + l.sales_matched, 0);

  // Accru matched
  let accruMatched = 0;
  for (const ml of matchedLevels) { accruMatched += ml.sales_matched; ml.accru_matched = accruMatched; }

  // Slippage
  const expectedThb = volUsed * bestBid;
  const diff = volReceived - expectedThb;
  const slippage = expectedThb > 0 ? diff / expectedThb : 0;

  // Safety line
  const safety = calculateSafetyLine(levels, bestBid, threshold);

  return {
    best_bid: bestBid, worst_bid: worstBid, total_amount: totalAmount,
    total_thb: levels[levels.length - 1].accru_thb,
    vol_used: volUsed, vol_received: volReceived, diff, slippage,
    slippage_pct: (slippage * 100).toFixed(3) + '%',
    threshold, threshold_breached: slippage < threshold,
    safety, levels: matchedLevels,
  };
}

function fillOrders(levels, volTarget) {
  const matched = [];
  let remaining = volTarget;
  for (const level of levels) {
    if (remaining <= 0) { matched.push({ ...level, amount_match: 0, sales_matched: 0 }); continue; }
    const matchAmount = Math.min(remaining, level.amount);
    const matchThb = matchAmount * level.price;
    remaining -= matchAmount;
    matched.push({ ...level, amount_match: matchAmount, sales_matched: matchThb });
  }
  return matched;
}

function calculateSafetyLine(levels, bestBid, threshold) {
  if (bestBid <= 0 || !levels.length) return { safe_vol: 0, safe_thb: 0, crossed_at_level: -1, is_safe: false };

  let accruVol = 0, accruThb = 0;
  for (let i = 0; i < levels.length; i++) {
    const { price, amount } = levels[i];
    const nextVol = accruVol + amount;
    const nextThb = accruThb + amount * price;
    const expected = nextVol * bestBid;
    const slip = expected > 0 ? (nextThb - expected) / expected : 0;

    if (slip < threshold) {
      const t = threshold;
      const num = (1 + t) * accruVol * bestBid - accruThb;
      const den = price - (1 + t) * bestBid;
      const x = den !== 0 ? Math.max(0, Math.min(num / den, amount)) : 0;
      return { safe_vol: accruVol + x, safe_thb: accruThb + x * price, crossed_at_level: i, is_safe: false };
    }
    accruVol = nextVol;
    accruThb = nextThb;
  }
  return { safe_vol: accruVol, safe_thb: accruThb, crossed_at_level: -1, is_safe: true };
}

function emptyResult(threshold) {
  return {
    best_bid: 0, worst_bid: 0, total_amount: 0, total_thb: 0,
    vol_used: 0, vol_received: 0, diff: 0, slippage: 0,
    slippage_pct: 'N/A', threshold, threshold_breached: false,
    safety: { safe_vol: 0, safe_thb: 0, crossed_at_level: -1, is_safe: false },
    levels: [],
  };
}
